xdsl/transforms/stencil_shape_minimize.py

```diff
@@ -0,0 +1,73 @@
+from dataclasses import dataclass, field
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, func, stencil
+from xdsl.ir import Attribute
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    TypeConversionPattern,
+    attr_type_rewrite_pattern,
+    op_type_rewrite_pattern,
+)
+
+
+@dataclass
+class ShapeAnalysis(TypeConversionPattern):
+    seen: set[stencil.TempType[Attribute]] = field(default_factory=set)
+
+    @attr_type_rewrite_pattern
+    def convert_type(self, typ: stencil.TempType[Attribute], /) -> Attribute | None:
+        self.seen.add(typ)
+
+
+@dataclass
+class ShapeMinimisation(TypeConversionPattern):
+    shape: stencil.StencilBoundsAttr | None = None
+
+    @attr_type_rewrite_pattern
+    def convert_type(self, typ: stencil.FieldType[Attribute], /) -> Attribute | None:
+        if typ.bounds != self.shape and self.shape:
```

Comment:
```suggestion
    shape: stencil.StencilBoundsAttr

    @attr_type_rewrite_pattern
    def convert_type(self, typ: stencil.FieldType[Attribute], /) -> Attribute | None:
        if typ.bounds != self.shape:
```


Since we always pass the shape on construction anyway

Reply:
That unfortunately doesn't seem to work with `dataclass`

Reply:
Ohh, `TypeConversionPattern` has default values, I didn't realise that.